src/main/java/org/dcsa/ebl/service/impl/TransportDocumentTOServiceImpl.java

```diff
@@ -1,28 +1,207 @@
+    private final ChargeService chargeService;
+    private final ClauseService clauseService;
+    private final LocationService locationService;
+    private final BookingService bookingService;
+    private final ShipmentService shipmentService;
+    private final ShipmentTransportService shipmentTransportService;
 
     @Override
-    public TransportDocumentTORepository getRepository() {
-        return transportDocumentTORepository;
+    public Mono<TransportDocumentTO> create(TransportDocumentTO transportDocumentTO) {
+        TransportDocument transportDocument = MappingUtil.instanceFrom(
+                transportDocumentTO,
+                TransportDocument::new,
+                AbstractTransportDocument.class
+        );
+        if (transportDocumentTO.getShippingInstruction() != null) {
+            return Mono.error(new CreateException("ShippingInstruction object cannot be included when creating a TransportDocument"));
+        } else {
+            return transportDocumentService.create(transportDocument)
+                    .flatMap(td -> {
+                        transportDocumentTO.setId(td.getId());
+                        return Flux.concat(
+                                shippingInstructionTOService.findById(transportDocument.getShippingInstructionID())
+                                        .doOnNext(shippingInstruction ->
+                                                transportDocumentTO.setShippingInstruction(shippingInstruction)
+                                        )
+                                        .thenReturn(transportDocumentTO),
+                                createCharges(transportDocumentTO),
+                                createClauses(transportDocumentTO),
+                                createTransportPlan(transportDocumentTO)
+                        )
+                        .then(Mono.just(transportDocumentTO));
+                    });
+        }
+    }
+
+    private Flux<Charge> createCharges(TransportDocumentTO transportDocumentTO) {
+        List<ChargeTO> chargeTOs = transportDocumentTO.getCharges();
+        if (chargeTOs == null || chargeTOs.isEmpty()) {
+            transportDocumentTO.setCharges(Collections.emptyList());
+            return Flux.empty();
+        } else {
+            List<Charge> charges = new ArrayList<>(chargeTOs.size());
+            chargeTOs.stream().forEach(chargeTO -> {
+                // Insert TransportDocumentID on all Charges
+                chargeTO.setTransportDocumentID(transportDocumentTO.getId());
+                // Create a Charge object for all ChargeTOs
+                Charge charge = MappingUtil.instanceFrom(
+                        chargeTO,
+                        Charge::new,
+                        AbstractCharge.class
+                );
+                charges.add(charge);
+            });
+
+            // Save all Charges in one bulk
+            return chargeService.createAll(charges);
```

Comment:
Can be rewritten as:
```
  Flux.fromIterable(chargeTOs)
  .map(cargeTO -> {
                // Insert TransportDocumentID on all Charges
                chargeTO.setTransportDocumentID(transportDocumentTO.getId());
                // Create a Charge object for all ChargeTOs
                Charge charge = MappingUtil.instanceFrom(
                        chargeTO,
                        Charge::new,
                        AbstractCharge.class
                );
                return charge;
  })
  .buffer(Util.<...>)
  .concatMap(cargeService::createAll)
```

Reply:
(deliberately not a suggestion because it might need a few finishing touches that the IDE can detect)